oshmem/mca/scoll/ucc/scoll_ucc_module.c

```diff
@@ -0,0 +1,357 @@
+                UCC_TEAM_PARAM_FIELD_EP_RANGE |
+                UCC_TEAM_PARAM_FIELD_OOB, 
+        .oob = {
+            .allgather    = oob_allgather,
+            .req_test     = oob_allgather_test,
+            .req_free     = oob_allgather_free,
+            .coll_info    = (void *) osh_group,
+            .participants = osh_group->proc_count, 
+        },
+        .ep       = osh_group->my_pe,
+        .ep_range = UCC_COLLECTIVE_EP_RANGE_CONTIG,
+    };
+
+    if (OSHMEM_SUCCESS != mca_scoll_ucc_save_coll_handlers(module, osh_group)) {
+        UCC_ERROR("UCC module enable failed");
+        /* There are no modules available */
+        opal_show_help("help-oshmem-scoll-ucc.txt",
+                       "module_enable:fatal", true,
+    	       		   "UCC module enable failed - aborting to prevent inconsistent application state");
+
+        return OSHMEM_ERROR;
+    }
+
+    if (UCC_OK != ucc_team_create_post(&cm->ucc_context, 1, 
+                                       &team_params, &ucc_module->ucc_team)) {
+        UCC_ERROR("ucc_team_create_post failed");
+    }
+
+    while (UCC_INPROGRESS == (status = ucc_team_create_test(ucc_module->ucc_team))) {
+        opal_progress();
+    }
+
+    if (UCC_OK != status) {
+        UCC_ERROR("ucc_team_create_test failed");
+        goto err;
+    }
+
+    return OSHMEM_SUCCESS;
+
+err:
+    ucc_module->ucc_team = NULL;
+    cm->ucc_enable = 0;
+    opal_progress_unregister(mca_scoll_ucc_progress);
+    return OSHMEM_ERROR;
+}
+
+#define SET_SCOLL_PTR(_module, _COLL, _coll) do {                       \
+    _module->super.scoll_  ## _coll = NULL;                             \
+    if ((mca_scoll_ucc_component.ucc_lib_attr.coll_types &              \
+         UCC_COLL_TYPE_ ## _COLL)) {                                    \
+        if (mca_scoll_ucc_component.cts_requested &                     \
+            UCC_COLL_TYPE_ ## _COLL) {                                  \
+            _module->super.scoll_ ## _coll  = mca_scoll_ucc_  ## _coll; \
+        }                                                               \
+    }                                                                   \
+} while(0)
+
+/*
+ * Invoked when there's a new communicator/group that has been created.
+ * Look at the communicator and decide which set of functions and
+ * priority we want to return.
+ */
+mca_scoll_base_module_t *
+mca_scoll_ucc_comm_query(oshmem_group_t *osh_group, int *priority)
+{
+    mca_scoll_base_module_t *module;
+    mca_scoll_ucc_module_t  *ucc_module;
+    *priority = 0;
+    mca_scoll_ucc_component_t *cm;
+    cm = &mca_scoll_ucc_component;
+
+    if (!cm->ucc_enable){
+        return NULL;
+    }
+
+    if ((osh_group->proc_count < 2) || (osh_group->proc_count < cm->ucc_np)) {
+        return NULL;
+    }
+    OPAL_TIMING_ENV_INIT(comm_query);
+
+    if (!cm->libucc_initialized) {
+        if (OSHMEM_SUCCESS != mca_scoll_ucc_init_ctx(osh_group)) {
```

Comment:
is the first group always world group, or is there will be the first group that the collective is called upon?

Reply:
Yes. I believe the first group should always be the world group.